openflow15/action.go

```diff
@@ -0,0 +1,605 @@
+	a := new(ActionPush)
+	a.Type = ActionType_PushVlan
+	a.Length = a.Len()
+	a.EtherType = etherType
+	return a
+}
+
+func NewActionPushMpls(etherType uint16) *ActionPush {
+	a := new(ActionPush)
+	a.Type = ActionType_PushMpls
+	a.Length = a.Len()
+	a.EtherType = etherType
+	return a
+}
+
+func (a *ActionPush) Len() (n uint16) {
+	return a.ActionHeader.Len() + 4
+}
+
+func (a *ActionPush) MarshalBinary() (data []byte, err error) {
+	data, err = a.ActionHeader.MarshalBinary()
+
+	bytes := make([]byte, 4)
+	binary.BigEndian.PutUint16(bytes[0:], a.EtherType)
+
+	data = append(data, bytes...)
+	return
+}
+
+func (a *ActionPush) UnmarshalBinary(data []byte) error {
+	a.ActionHeader.UnmarshalBinary(data[:4])
+	a.EtherType = binary.BigEndian.Uint16(data[4:])
+	return nil
+}
+
+type ActionPopVlan struct {
+	ActionHeader
+	pad []byte // 4bytes
+}
+
+func NewActionPopVlan() *ActionPopVlan {
+	act := new(ActionPopVlan)
+	act.Type = ActionType_PopVlan
+	act.Length = act.Len()
+
+	return act
+}
+
+func (a *ActionPopVlan) Len() (n uint16) {
+	return a.ActionHeader.Len() + 4
+}
+
+func (a *ActionPopVlan) MarshalBinary() (data []byte, err error) {
+	data, err = a.ActionHeader.MarshalBinary()
+
+	// Padding
+	bytes := make([]byte, 4)
+
+	data = append(data, bytes...)
+	return
+}
+
+func (a *ActionPopVlan) UnmarshalBinary(data []byte) error {
+	a.ActionHeader.UnmarshalBinary(data[:4])
+	return nil
+}
+
+type ActionPopMpls struct {
+	ActionHeader
+	EtherType uint16
+	pad       []byte // 2bytes
+}
+
+func NewActionPopMpls(etherType uint16) *ActionPopMpls {
+	act := new(ActionPopMpls)
+	act.Type = ActionType_PopMpls
+	act.EtherType = etherType
+	act.Length = act.Len()
+
+	return act
+}
+
+func (a *ActionPopMpls) Len() (n uint16) {
+	return a.ActionHeader.Len() + 4
+}
+
+func (a *ActionPopMpls) MarshalBinary() (data []byte, err error) {
+	data, err = a.ActionHeader.MarshalBinary()
+
+	// Padding
+	bytes := make([]byte, 4)
+	binary.BigEndian.PutUint16(bytes[0:], a.EtherType)
+
+	data = append(data, bytes...)
+	return
+}
+
+func (a *ActionPopMpls) UnmarshalBinary(data []byte) error {
+	a.ActionHeader.UnmarshalBinary(data[:4])
+	a.EtherType = binary.BigEndian.Uint16(data[4:])
+	return nil
+}
+
+type ActionSetField struct {
+	ActionHeader
+	Field MatchField
+}
+
+func NewActionSetField(field MatchField) *ActionSetField {
+	a := new(ActionSetField)
+	a.Type = ActionType_SetField
+	a.Field = field
+	a.Length = a.Len()
+	return a
+}
+
+func (a *ActionSetField) Len() (n uint16) {
+	n = a.ActionHeader.Len() + a.Field.Len()
+	// Round it to closest multiple of 8
+	n = ((n + 7) / 8) * 8
+
+	return
+}
+
+func (a *ActionSetField) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, int(a.Len()))
+	n := 0
+	b, err := a.ActionHeader.MarshalBinary()
+	if err != nil {
+		return
+	}
+	copy(data, b)
+	n += int(a.ActionHeader.Len())
+
+	b, err = a.Field.MarshalBinary()
+	copy(data[n:], b)
+
+	return
+}
+func (a *ActionSetField) UnmarshalBinary(data []byte) error {
```

Comment:
Could you insert an empty line between methods?

Reply:
Sure. Would do.